File: extension/core/src/main/java/io/holunda/camunda/bpm/data/adapter/ReadAdapter.java
package io.holunda.camunda.bpm.data.adapter;

import java.util.Optional;

/**
 * Adapter to read variables.
 *
 * @param <T> type of value.
 */
public interface ReadAdapter<T> {

  /**
   * Reads a variable.
   *
   * @return value.
   * @exception VariableNotFoundException if the required variable is missing or can't be read.
   */
  T get();

  /**
   * Reads a variable and returns a value if exists or an empty.
   *
   * @return optional.
   */
  Optional<T> getOptional();

  /**
   * Reads a local variable.
   *
   * @return value.
   * @exception VariableNotFoundException if the required variable is missing or can't be read.
   */
  T getLocal();

  /**
   * Reads a local variable and returns a value if exists or an empty.
   *
   * @return optional.
   */
  Optional<T> getLocalOptional();

  /**
   * Reads a variable and returns a value if exists or default.
   *
   * @param defaultValue the default value if the variable is not set
   * @return value or default
   */
  default T getOrDefault(T defaultValue) {
    return getOptional().orElse(defaultValue);
  }

  /**
   * Reads a local variable and returns a value if exists or default.
   *
   * @param defaultValue the default value if the variable is not set
   * @return value or default
   */
  default T getLocalOrDefault(T defaultValue) {
    return getLocalOptional().orElse(defaultValue);
  }

  /**
   * Reads a variable and returns a value if exists or null.
   *
   * @return value or <code>null</code>>
   */
  default T getOrNull() {
    return getOptional().orElse(null);
  }

  /**
   * Reads a local variable and returns a value if exists or null.
   *
   * @return value or <code>null</code>>
   */
  default T getLocalOrNull() {
    return getLocalOptional().orElse(null);
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/adapter/ReadAdapter.kt
package io.holunda.camunda.bpm.data.adapter

import java.util.*

/**
 * Adapter to read variables.
 *
 * @param [T] type of value.
 */
interface ReadAdapter<T> {
  /**
   * Reads a variable.
   *
   * @return value.
   * @exception VariableNotFoundException if the required variable is missing or can't be read.
   */
  fun get(): T

  /**
   * Reads a variable and returns a value if exists or an empty.
   *
   * @return optional.
   */

  fun getOptional(): Optional<T>

  /**
   * Reads a local variable.
   *
   * @return value.
   * @exception VariableNotFoundException if the required variable is missing or can't be read.
   */
  fun getLocal(): T

  /**
   * Reads a local variable and returns a value if exists or an empty.
   *
   * @return optional.
   */
  fun getLocalOptional(): Optional<T>

  /**
   * Reads a variable and returns a value if exists or default.
   *
   * @param defaultValue the default value if the variable is not set
   * @return value or default
   */
  fun getOrDefault(defaultValue: T): T {
    return getOptional().orElse(defaultValue)
  }

  /**
   * Reads a local variable and returns a value if exists or default.
   *
   * @param defaultValue the default value if the variable is not set
   * @return value or default
   */
  fun getLocalOrDefault(defaultValue: T): T {
    return getLocalOptional().orElse(defaultValue)
  }

  /**
   * Reads a variable and returns a value if exists or null.
   *
   * @return value or `null`
   */
  fun getOrNull(): T? {
    return getOptional().orElse(null)
  }

  /**
   * Reads a local variable and returns a value if exists or null.
   *
   * @return value or `null`
   */
  fun getLocalOrNull(): T? {
    return getLocalOptional().orElse(null)
  }
}
